sematic/abstract_future.py

```diff
@@ -65,6 +65,56 @@ def terminal_states(cls) -> FrozenSet["FutureState"]:
     def terminal_state_strings(cls) -> FrozenSet[str]:
         return _TERMINAL_STATE_STRINGS
 
+    @classmethod
+    def is_allowed_transition(
+        cls,
+        from_status: Optional[Union["FutureState", str]],
+        to_status: Union["FutureState", str],
+    ) -> bool:
+        if isinstance(from_status, str):
+            from_status = FutureState[from_status]
+        if isinstance(to_status, str):
+            to_status = FutureState[to_status]
+        return to_status in _ALLOWED_TRANSITIONS[from_status]
+
+
+_ALLOWED_TRANSITIONS = {
+    None: frozenset({FutureState.CREATED}),
+    FutureState.CREATED: frozenset(
+        {
+            FutureState.SCHEDULED,
+            # Not all that uncommon: if run is downstream from something that
+            # is canceled/failed, it goes from CREATED -> CANCELLED
+            FutureState.CANCELED,
+        }
+    ),
+    FutureState.SCHEDULED: frozenset(
+        {
+            FutureState.RAN,
+            FutureState.RETRYING,
+            FutureState.CANCELED,
+            FutureState.RESOLVED,
+            # TODO: remove (See https://github.com/sematic-ai/sematic/issues/609)
+            FutureState.FAILED,
```

Comment:
What if the func does not have retries configured? Then `SCHEDULED -> FAILED` should be kept as a valid transition.

Reply:
I think the resolver should be the one to decide what happens. The worker can just "blindly" go from `SCHEDULED -> RETRYING` and the resolver can go from `RETRYING` -> `FAILED` if the run doesn't have any retries allotted. Though that's an implementation detail for a separate PR anyway.